### DISCUSSION TITLE
[Deployment] BLS signature verification on ZkSync Era

### DISCUSSION THREAD

**Discussion by harsh-supra** [zkSync-Community-Hub/zksync-developers#167] [Deployments] [answered]
### Environment

Testnet

### zkSolc Version

0.4.2

### zksync-web3 Version

0.14.3

### Hardhat.config.ts

import { HardhatUserConfig } from "hardhat/config";

import "@matterlabs/hardhat-zksync-deploy";
import "@matterlabs/hardhat-zksync-solc";

import "@matterlabs/hardhat-zksync-verify";

// dynamically changes endpoints for local tests
export const zkSyncTestnet =
  process.env.NODE_ENV == "test"
    ? {
        url: "http://127.0.0.1:8011",
        ethNetwork: "http://127.0.0.1:8045",
        zksync: true,
      }
    : {
        url: "https://damp-quick-shadow.zksync-testnet.quiknode.pro/6a80da56f388295cbfe3686d771e0ab14c52379f/",
        ethNetwork: "goerli",
        zksync: true,
        // contract verification endpoint
        verifyURL:
          "https://zksync2-testnet-explorer.zksync.dev/contract_verification",
      };

const config: HardhatUserConfig = {
  zksolc: {
    version: "latest",
    settings: {
        optimizer: {
            enabled: true,
            mode: "z",
            runs: 200,
        },
    },
  },
  defaultNetwork: "zkSyncTestnet",
  networks: {
    hardhat: {
      zksync: false,
    },
    zkSyncTestnet,
  },
  solidity: {
    version: "0.8.19",
  },
};

export default config;

### Deployment Script (WITHOUT PRIVATE KEY)

import { Provider, Wallet } from "zksync-web3";
import * as ethers from "ethers";
import { HardhatRuntimeEnvironment } from "hardhat/types";

// load env file
import dotenv from "dotenv";
dotenv.config();

// load contract artifact. Make sure to compile first!
import * as ContractArtifact from "../artifacts-zk/contracts/Verify.sol/Verify.json";

const PRIVATE_KEY = process.env.WALLET_PRIVATE_KEY || "";

if (!PRIVATE_KEY)
    throw "⛔️ Private key not detected! Add it to the .env file!";

// Address of the contract on zksync testnet
const CONTRACT_ADDRESS = "0x531a90d73a5eA8735b0b0350851663F288BBD4c8";

if (!CONTRACT_ADDRESS) throw "⛔️ Contract address not provided";

// An example of a deploy script that will deploy and call a simple contract.
export default async function (hre: HardhatRuntimeEnvironment) {
    console.log(`Running script to interact with contract ${CONTRACT_ADDRESS}`);

    // Initialize the provider.
    // @ts-ignore
    const provider = new Provider(hre.userConfig.networks?.zkSyncTestnet?.url);
    const signer = new Wallet(PRIVATE_KEY, provider);

    // Initialize contract instance
    const contract = new ethers.Contract(
        CONTRACT_ADDRESS,
        ContractArtifact.abi,
        signer
    );

    // Read message from contract
    console.log(`The message is ${await contract.verify_bls()}`);

### Contract Code

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.4;

import "./BLS.sol";

contract Verify {
    bytes32 domain;
    uint256[4] publicKey;
    uint256[2] signature;
    uint64 gasCost = 113133;

    constructor() {
        domain = 0x0000000000000000000000000000000000000000000000000000000000000020;
        publicKey[0] =  1877680754511875309899085821046020641041516699522550968201931210511122361188;
        publicKey[1] =  1879687745237862349771417085220368195630510774060410176566704734657946401647;
        publicKey[2] =  10177664824327229270631241062558466194853353905576267792570130720130119743401;
        publicKey[3] =  2617838070911723228053200997531205923494421078683439118196637157934995837361;

        signature[0] = 3789870118542016974105699138161781008918636358438246688710078193373583696417;
        signature[1] = 20766815278548398344257910495536643911108480586016736348038320126600861929561;
    }

    function verify_bls() public view returns (bool,bool) {
        bytes32 message = 0x321580d53f250a51ed527f5f7856bdd2ecbbf19f722ee6acc1804f63462375f3;
        bool callSuccess;
        bool checkSuccess;
        (checkSuccess, callSuccess) = BLS.verifySingle(signature, publicKey, BLS.hashToPoint(domain, abi.encode(message)), gasCost);
//        require(callSuccess, "Incorrect Publickey or Signature Points");
//        require(checkSuccess, "Incorrect Input Message");
        return (checkSuccess, callSuccess);
    }
}

### Does this work on other EVMs? (If yes, please list at least 1 of them)

Yes, Ethereum , Polygon, Arbitrum, Mantle

### Description of What Your Contract Does

Verifies the BLS signature using the 0x8 BLS precompile

**Comment by Romsters** [comment, marked as answer]
Hi @harsh-supra , the reason it doesn't work is that you probably use a precompiled contract [alt_bn128](https://eips.ethereum.org/EIPS/eip-197) which is not yet supported on zkSync Era. Some of the precompiles will be released in a coming weeks, the alt_bn128 in a coming months.

**Reply by harsh-supra** [reply]
Ok, Noted
Thanks for the response @Romsters

**Reply by anshuman-supraoracles** [reply]
Hi @Romsters Do zKSyn Era started supporting alt_bn128?

**Reply by bxpana** [reply]
When we have support for different precompiles we'll announce on https://twitter.com/zkSyncDevs

**Comment by anshuman-supraoracles** [comment]
Hi @Romsters Do zKSyn Era started supporting alt_bn128?

**Comment by jakeaukuso** [comment]
It seems like you're encountering an issue when trying to verify contract coverage with zkSync. The error message suggests a problem with the bytecode generation during compilation. This could be due to compatibility issues between the zkSync environment and the coverage tool being used. You might want to check for any updates or patches specifically addressing coverage testing with zkSync, or consider alternative approaches for verifying contract coverage in a zkSync environment.

**Comment by kakarukjoseph** [comment]
It seems that the issue might be related to the configuration of your zk-sync testnet provider URL. Please ensure that the URL is correct and accessible. Additionally, double-check the contract address and make sure it is deployed on the zk-sync testnet. If the issue persists, consider reaching out to the zk-sync development community for further assistance.

**Comment by jonnybutlein** [comment]
// This script is designed to deploy and interact with a Solidity contract that verifies BLS signatures using the 0x8 BLS precompile on zkSync testnet. It configures the Hardhat environment, specifies the zkSync testnet network, and initializes the deployment and interaction process with the contract.

**Comment by dyksterhouset** [comment]
Failed to run coverage for zkSync contracts due to compiler error "write EPIPE".
